pkg/client/testing/core/fixture.go

```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
-			arr, ok := o.types[itemKind]
-			if !ok {
-				return empty, nil
-			}
-			out, err := o.scheme.New(kind)
-			if err != nil {
-				return nilValue, err
-			}
-			if err := meta.SetList(out, arr); err != nil {
-				return nilValue, err
-			}
-			if out, err = o.scheme.Copy(out); err != nil {
-				return nilValue, err
-			}
-			return out, nil
-		}
-		return nilValue, errors.NewNotFound(unversioned.GroupResource{Group: kind.Group, Resource: kind.Kind}, name)
+		return list, nil
 	}
 
-	index := o.last[kind.Kind]
-	if index >= len(arr) {
-		index = len(arr) - 1
+	matchingObjs, err := filterByNamespaceAndName(objs, ns, "")
+	if err != nil {
+		return nil, err
 	}
-	if index < 0 {
-		return nilValue, errors.NewNotFound(unversioned.GroupResource{Group: kind.Group, Resource: kind.Kind}, name)
+	if len(matchingObjs) == 0 {
+		return list, nil
 	}
-	out, err := o.scheme.Copy(arr[index])
+	if err := meta.SetList(list, matchingObjs); err != nil {
+		return nil, err
+	}
+	if list, err = t.scheme.Copy(list); err != nil {
+		return nil, err
+	}
+	return list, nil
+}
+
+func (t *tracker) Get(gvk unversioned.GroupVersionKind, ns, name string) (runtime.Object, error) {
+	obj, err := t.scheme.New(gvk)
 	if err != nil {
-		return nilValue, err
+		return nil, err
 	}
-	o.last[kind.Kind] = index + 1
+	errNotFound := errors.NewNotFound(unversioned.GroupResource{Group: gvk.Group, Resource: gvk.Kind}, name)
+
+	t.lock.RLock()
+	defer t.lock.RUnlock()
 
-	if status, ok := out.(*unversioned.Status); ok {
+	objs, ok := t.objects[gvk]
+	if !ok {
+		return nil, errNotFound
+	}
+
+	matchingObjs, err := filterByNamespaceAndName(objs, ns, name)
+	if err != nil {
+		return nil, err
+	}
+	if len(matchingObjs) == 0 {
+		return nil, errNotFound
+	}
+
+	// Only one object should match in the tracker if it works
+	// correctly, as Add/Update methods enforce kind/namespace/name
+	// uniqueness.
+	obj, err = t.scheme.Copy(matchingObjs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	if status, ok := obj.(*unversioned.Status); ok {
 		if status.Details != nil {
-			status.Details.Kind = kind.Kind
+			status.Details.Kind = gvk.Kind
 		}
 		if status.Status != unversioned.StatusSuccess {
-			return nilValue, &errors.StatusError{ErrStatus: *status}
+			return nil, &errors.StatusError{ErrStatus: *status}
 		}
 	}
 
-	return out, nil
+	return obj, nil
+}
+
+func (t *tracker) Add(obj runtime.Object) error {
+	return t.add(obj, false)
 }
 
-func (o objects) Add(obj runtime.Object) error {
-	gvks, _, err := o.scheme.ObjectKinds(obj)
+func (t *tracker) Update(obj runtime.Object) error {
+	return t.add(obj, true)
+}
+
+func (t *tracker) add(obj runtime.Object, replaceExisting bool) error {
+	if meta.IsListType(obj) {
+		return t.addList(obj, replaceExisting)
+	}
+
+	gvks, _, err := t.scheme.ObjectKinds(obj)
 	if err != nil {
 		return err
 	}
-	kind := gvks[0].Kind
+	if len(gvks) == 0 {
+		return fmt.Errorf("no registered kinds for %v", obj)
+	}
+	gvk := gvks[0]
```

Comment:
Seems like you should loop through and add it to all of them.  Hitting an error would fail the call and leave this instance in an unknown state, that's ok.